Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -803,42 +803,37 @@ void Device::draw_primitives(PrimitiveType primitive_type, FloatMatrix4x4 const&
                     if (!light.is_enabled)
                         continue;
 
-                    // We need to save the length here because the attenuation factor requires a non
-                    // normalized vector!
-                    auto sgi_arrow_operator = [](FloatVector4 const& p1, FloatVector4 const& p2, float& saved_length) {
+                    // We need to save the length here because the attenuation factor requires a non-normalized vector!
+                    auto sgi_arrow_operator = [](FloatVector4 const& p1, FloatVector4 const& p2, float& output_length) {
                         if ((p1.w() != 0.0f) && (p2.w() == 0.0f)) {
-                            saved_length = p2.length();
-                            return (p2 / saved_length).xyz();
+                            output_length = p2.length();
+                            return p2.xyz().normalized();
```

Comment:
Since we already have the length, we don't need to call normalize which will calculate the length *again*. We can just divide the vector by the length. This will save another `sqrt` call